R Markdown Enhancements (Knit Manager)

ElianHugh opened the pull request:
Enhances rendering R Markdown documents. Specifically: 

- background knitting (with progress bar)
- set knit working directory
- Smart knit button, use custom knit functions + rmarkdown::render_site where appropriate


## Screenshots

![image](https://user-images.githubusercontent.com/60372411/130966958-7f6df1b0-363e-4681-bffe-c93d6b505712.png)

![image](https://user-images.githubusercontent.com/60372411/130970527-01e8d756-1616-4170-a6cb-f82c148749e4.png)

![image](https://user-images.githubusercontent.com/60372411/130967133-bc98cba5-80c5-42e1-bfbe-913d7bf20b15.png)

### "Smart" Knit Button

#### Custom Knit

![image](https://user-images.githubusercontent.com/60372411/130972698-4a91d6b6-6fa3-4d80-9909-411fd9b5a03e.png)

![image](https://user-images.githubusercontent.com/60372411/130972728-222a4d94-7636-41dc-95a2-49390811892e.png)

#### R Markdown Site

![image](https://user-images.githubusercontent.com/60372411/130973068-093e49aa-0511-49ae-8c22-89627bcda204.png)

![image](https://user-images.githubusercontent.com/60372411/130973142-944fff7c-3ff8-492c-9d0e-c7c68f41860d.png)

renkun-ken commented:
Always fail to start debugging. The following error occurs on extension activation:

```
Activating extension 'Ikuyadeu.r' failed: rmarkdown.RMarkdownPreviewManager is not a constructor.
```

renkun-ken commented on src/rmarkdown/manager.ts:
```diff
@@ -0,0 +1,146 @@
+	protected busyUriStore: Set<string> = new Set<string>();
+
+	public async init(): Promise<void> {
+		this.rPath = await util.getRpath(true);
```
If `rPath` is determined in `init()`, then chaning R paths in settings will not have effect before restart. Maybe we should call `util.getRpath()` each time we use it?

ElianHugh commented:
> 
> 
> Always fail to start debugging. The following error occurs on extension activation:
> 
> ```
> Activating extension 'Ikuyadeu.r' failed: rmarkdown.RMarkdownPreviewManager is not a constructor.
> ```

I suspect this is some kind of bug with TypeScript, as I had this happen until I deleted my `out` folder. It could also be the version of TS installed? I've updated to 4.3.5. In any case, I can't see why a wildstar import would be invalid for classes, but explicit exports would be ok

renkun-ken commented:
> as I had this happen until I deleted my out folder

This actually works nicely! Thanks!

renkun-ken commented on src/rmarkdown/knit.ts:
```diff
@@ -149,6 +149,7 @@ export class RMarkdownKnitManager extends RMarkdownManager {
 
 	// alters the working directory for evaluating chunks
 	public setKnitDir(): void {
+		const currentDocumentWorkspace = vscode.workspace.getWorkspaceFolder(vscode.window.activeTextEditor.document.uri);
```
If the document is untitled but set to `rmd` language, then the workspace folder is `undefined`, if I remember correctly.

ElianHugh replied:
Ah, yes that would cause some trouble. Given that a document can't be knit without being saved, I think it would make sense to have the following fallbacks:

Workspace folder > document folder > NULL

renkun-ken commented on src/rmarkdown/knit.ts:
```diff
@@ -149,20 +149,26 @@ export class RMarkdownKnitManager extends RMarkdownManager {
 	// alters the working directory for evaluating chunks
 	public setKnitDir(): void {
-		const currentDocumentWorkspace = vscode.workspace.getWorkspaceFolder(vscode.window.activeTextEditor.document.uri);
+		let currentDocumentWorkspacePath: string = null;
```
I should have made it more clear: untitled document and document outside any workspace folder will return `undefined` workspace folder. We should handle both cases separately.

renkun-ken replied:
Maybe we could provide QuickPickItem accordingly:

* If it has a workspace folder, then add a QuickPickItem for it.
* If it has a filesystem folder, then add a QuickPickItem for it.

renkun-ken reviewed (approved):
Works nicely!

LGTM

ManuelHentschel commented:
On windows, if I set the Knit directory to the document directory, I get an error:
```
[VSC-R] test.Rmd process started
Erreur : '\\' inattendu(e) dans "knitr::opts_knit[['set']](root.dir = c:\\"
Ex�cution arr�t�e

[VSC-R] test.Rmd process exited with exit code 1
```
In the workspace root this error does not occur.

ElianHugh commented:
> On windows, if I set the Knit directory to the document directory, I get an error:
> 
> ```
> [VSC-R] test.Rmd process started
> Erreur : '\\' inattendu(e) dans "knitr::opts_knit[['set']](root.dir = c:\\"
> Ex�cution arr�t�e
> 
> [VSC-R] test.Rmd process exited with exit code 1
> ```
> 
> In the workspace root this error does not occur.

Argh, the path differences are the bane of my existence!! Thanks for catching that, *hopefully* the most recent commit fixes this?

ManuelHentschel commented on package.json:
```diff
@@ -1262,7 +1299,37 @@
+        },
+        "r.rmarkdown.knit.defaults.knitWorkingDirectory": {
+          "type": "string",
+          "default": "document",
```
This should probably be `document directory`

ManuelHentschel commented on src/rmarkdown/knit.ts:
```diff
@@ -0,0 +1,250 @@
+import { RMarkdownManager, KnitWorkingDirectory } from './manager';
+import { runTextInTerm } from '../rTerminal';
+
+export let knitDir: string = util.config().get<string>('rmarkdown.knit.defaults.knitWorkingDirectory') ?? undefined;
```
Preferrably this should be typed as `KnitWorkingDirectory` (to make clear that it does not contain e.g. a literal path)

ManuelHentschel commented on src/rmarkdown/knit.ts:
```diff
@@ -0,0 +1,250 @@
+				{
+					title: 'Set knit working directory',
+					canPickMany: false,
+					onDidSelectItem: (item: IKnitQuickPickItem) => {
+						knitDir = item.value;
+					}
```
`onDidSelectItem` triggers everytime you change the selection in the quickpick, e.g. using the arrow keys.
This leads to wrong behaviour, e.g. when the user highlights an option and then presses escape.

A better way is to use the return value of `showQuickPick` (which needs to be checked for `undefined` in case the user presses escape)

ManuelHentschel commented on src/rmarkdown/knit.ts:
```diff
@@ -0,0 +1,250 @@
+		} else {
+			void vscode.window.showInformationMessage('Cannot set knit directory for untitled documents.');
+		}
+
```
If the user changed the knit directory, it might be useful to trigger a refresh here

ManuelHentschel commented on src/rmarkdown/index.ts:
```diff
@@ -6,6 +6,10 @@ import {
 import { runChunksInTerm } from '../rTerminal';
 import { config } from '../util';
 
```
I can't add the comment at the corresponding lines (since it's not actually part of this PR), but I'd definitely prefer to use a `*` import for `vscode` here

ManuelHentschel commented:
> Argh, the path differences are the bane of my existence!! Thanks for catching that, _hopefully_ the most recent commit fixes this?

Can confirm that I'm not getting the error anymore

ManuelHentschel reviewed (approved):
I'm not very experienced with Rmd, but the simple examples I tried worked nicely.

Thanks for the nice work!
